### PR TITLE
utils: iso9660 and cloudinit has pycdlib as requirements

### PR DESCRIPTION
**Pull request by beraldoleal**
IMO, pycdlib should be an Avocado requirement. For more details, please, visit each commit description.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -158,4 +158,4 @@ def get_long_description():
           zip_safe=False,
           test_suite='selftests',
           python_requires='>=3.4',
-          install_requires=['setuptools'])
+          install_requires=['setuptools', 'pycdlib==1.6.0'])
```
Comment (willianrampazzo) [inline setup.py:161]: I understand the reason to add `pycdlib` as a requirement, but I wonder if this is the right thing. I mean, Avocado may depend on other packages in other modules, and following the logic here, we would need to make all of them a requirement.

The question I would make is _does it make sense for 100% of Avocado's users to have `pycdlib` installed in their environment?_ Maybe, most of them won't use it.

As you added in the previous commit, a better warning message is a conservative solution, which makes more sense for me at this moment.

Comment (beraldoleal) [reply setup.py:161]: I see your point, and your question was in mind too before submitting this. But IMO, having only the warning message is a **partial solution**.

We need to think about our user's experience and usability. Asking users to install a dependency IMO isn't the best approach to attract new users.

I would like to make a different question:  _as a tool that is defined as "a set of tools and libraries to help with automated testing", do you think that makes sense to install Avocado and force users to install other dependencies too to avoid error messages?_

IMO, this creates a terrible experience. Especially in this case where two libraries rely on this.

If this is a library that has just a few users (I don't think that is the case), then maybe would be a better move to migrate this to an "avocado-utils-extra" package.

Comment (willianrampazzo) [reply setup.py:161]: I agree with you here too, that is why I'm not against making `pycdlib` a dependency, nor in favor :)

In this case, I'm fine approving it.

**Review by willianrampazzo — APPROVED** [review]

File: setup.py (hunk 1 above)
Comment (clebergnu) [reply setup.py:161]: @beraldoleal I think your premise of usability is valid and correct, but we have to evaluate the other aspects of adding dependencies.  For instance, we worked hard to reduce Avocado do depend almost exclusively on Python (and almost "standard" setuptools) because some primary platforms that we want to target don't have all the fancy packages we could use as dependencies.  Then, diverging between upstream and packaged behavior is another inconvenience.  Also, all users having to install all dependencies, for which the capabilities are optional, is also a con IMO.

On the specific case of `cloudinit`, it's supposed to build upon `iso9660` which itself is supposed to be pluggable. There's no thing preventing us from adding a `mkisofs` based implementation of the `create` and `write` *capabilities*.  That's why the error message was generic and does not include `pycdlib` itself, even though it's the only capable backend at this point.  Adding to the previous point, if we add a `mkisofs` based write capability, then should we still install `pycdlib` every time we install Avocado?

So to summarize it, I'm against adding all dependencies that augment Avocado to all users (we've been there before), but I'm in favor of being more explicit about.  I'd try listing the backends that have the `create` and `write` capabilities, and telling users that a backend is not active because of missing software.  For `iso9660.ISO9660PyCDLib` it would obviously be the `pycdlib` Python module, but say, for `iso9660.MkIsoFs` it would be `/usr/bin/mkisofs`.

Another idea that strikes from this, is having an `avocado-utils capabilities` command.  It could list those optional features, and even install all possible features with a `avocado-utils install-all`.  Using software-manager we could attempt to install `/usr/bin/mkisofs`-like tools.

Thoughts?

Comment (beraldoleal) [reply setup.py:161]: @clebergnu thank you for your comments.

> For instance, we worked hard to reduce Avocado do depend almost exclusively on Python (and almost "standard" setuptools) because some primary platforms that we want to target don't have all the fancy packages we could use as dependencies.

Yes, I understand this, but if we would like to keep Avocado portable and Python only dependent, then maybe is time to rethink our distribution model. Because IMO it is hard to understand the following statements:  _Avocado works on some platforms. Avocado provides some features/utilities. But to have some features users have to install dependencies that are not available on all platforms._

> So to summarize it, I'm against adding all dependencies that augment Avocado to all users (we've been there before), but I'm in favor of being more explicit about. I'd try listing the backends that have the create and write capabilities, and telling users that a backend is not active because of missing software. 

IMO this is not ideal. For the reasons that I explained before. Having users to install dependencies by themselves is painfully and will reduce the Avocado's adoption, even if it is explicit at README/INSTALL instructions or when he/she is using the tool. If we have some features that we have multiple backends, we could choose at least one to make usable by default in order to deliver the feature.

But if the entire utility is optional and/or not so useful to most of our users, why we are keeping this as part of our core? Let's say, migrate this library to an "avocado-utils-extra" or "avocado-utils-iso" package? Yes, one more repo and one package to release but, Avocado is getting big and I'm missing what is the advantage of keeping libraries in our main package that aren't so useful for most of our users.

> 
Another idea that strikes from this, is having an avocado-utils capabilities command. It could list those optional features, and even install all possible features with an avocado-utils install-all. Using software-manager we could attempt to install /usr/bin/mkisofs-like tools.

I also don't like much the idea of having specific commands for utils here. Instead of that, I'm more in favor of having proper package dependency management. So just install "avocado-utils-iso" if you need iso stuff.

What I'm missing?

Comment (clebergnu) [reply setup.py:161]: > @clebergnu thank you for your comments.
> 
> > For instance, we worked hard to reduce Avocado do depend almost exclusively on Python (and almost "standard" setuptools) because some primary platforms that we want to target don't have all the fancy packages we could use as dependencies.
> 
> Yes, I understand this, but if we would like to keep Avocado portable and Python only dependent, then maybe is time to rethink our distribution model. Because IMO it is hard to understand the following statements: _Avocado works on some platforms. Avocado provides some features/utilities. But to have some features users have to install dependencies that are not available on all platforms._
> 

I agree we should rethink out distribution model, so much that we have discussed it (splitting `avocado.utils` out) and IMO we should move towards implementing it.

> > So to summarize it, I'm against adding all dependencies that augment Avocado to all users (we've been there before), but I'm in favor of being more explicit about. I'd try listing the backends that have the create and write capabilities, and telling users that a backend is not active because of missing software.
> 
> IMO this is not ideal. For the reasons that I explained before. Having users to install dependencies by themselves is painfully and will reduce the Avocado's adoption, even if it is explicit at README/INSTALL instructions or when he/she is using the tool. If we have some features that we have multiple backends, we could choose at least one to make usable by default in order to deliver the feature.
> 
> But if the entire utility is optional and/or not so useful to most of our users, why we are keeping this as part of our core? Let's say, migrate this library to an "avocado-utils-extra" or "avocado-utils-iso" package? Yes, one more repo and one package to release but, Avocado is getting big and I'm missing what is the advantage of keeping libraries in our main package that aren't so useful for most of our users.
> 

Yes, this is the trend we should follow. We have successfully followed this approach with the "optional plugins", which can and have extra dependencies while leaving the core Avocado simpler and more portable.  IMO it's time to follow suit with the utilities too.

> > 
> 
> Another idea that strikes from this, is having an avocado-utils capabilities command. It could list those optional features, and even install all possible features with an avocado-utils install-all. Using software-manager we could attempt to install /usr/bin/mkisofs-like tools.
> 
> I also don't like much the idea of having specific commands for utils here. Instead of that, I'm more in favor of having proper package dependency management. So just install "avocado-utils-iso" if you need iso stuff.
> 
> What I'm missing?

+1 for going *first* with a Python package dependency approach.  But, even a perfect *Python* package depedency deployment won't solve the `mkisofs` and `qemu-img` kind of scenarios so a helper command *may* be useful.  Time will tell.

Comment (beraldoleal) [reply setup.py:161]: Thanks for the clarification, @clebergnu. Good to know that we are on the same page.

I will keep this PR only with the "Partial solution" (warning message) and open an issue to move utils and optional plugins to other repos.

Comment (clebergnu) [reply setup.py:161]: OK, I'm waiting now for the changes you described here.

Comment (beraldoleal) [reply setup.py]: @clebergnu done.

**Review by clebergnu — APPROVED** [review]
LGTM, and CI is happy. Thanks!
